### PR TITLE
Added inheritance, special methods. Resolves #4.

### LINKED ISSUE #4: Add support for inheritance
Right now, the only methods, fields, etc. that will get documented are those directly on the class you're inspecting. It would be awesome if this library walked the inheritance tree to get all the upstream public methods and fields.

Doing this would mean that the docs for each class faithfully represent the true interface for that class.

### PR DESCRIPTION
**Pull request by jayqi**
Changes:
- Added support for class inheritance. 
    - Will recursively search for ancestor class methods and fields.
    - Added a line that contains a link to the parent class Roxygen page
- Renamed "member" to "field". The term "members" in OOP is a super-category that contains "variables" (called "fields" in R6) and "methods". See [one reference](https://www.quora.com/What-are-the-two-types-of-class-members-used-in-object-oriented-programming-languages), but can google more.
- Added a section "Special Methods" for special R6 methods like `finalize`, `print`, and `clone`. `clone` always exists, while the other two depend on people defining them. 
- Suppressed the warning that is otherwise unavoidable when getting the formals of a method that has none. 

Resolves #4

### CODE REVIEW

File: R/roxy_docsy.R [hunk 1]
```diff
@@ -86,64 +116,164 @@
...
+        if (!thisFuncName %in% names(aClass$public_methods)) {
+            next
+        }
+        
+        thisFuncDescript <- .describe_function(aClass$public_methods[[thisFuncName]], thisFuncName)
+        
+        # if the function is clone, put in standard documentation
+        if (thisFuncName == "clone") {
+            thisFuncDescript <- sub("~~DESCRIBE THE METHOD~~", CLONE_DESCRIBE, thisFuncDescript)
+            thisFuncDescript <- sub("~~DESCRIBE THIS PARAMETER ", CLONE_PARAM, thisFuncDescript)
```
Comment (jayqi) [inline R/roxy_docsy.R]:
```suggestion
            thisFuncDescript <- sub("~~DESCRIBE THIS PARAMETER~~", CLONE_PARAM, thisFuncDescript)
```
When #11 gets merged.

Comment (jameslamb) [reply R/roxy_docsy.R]: first time I've seen the new suggested-changes thing!

#11 is merged now, bw

File: .Rbuildignore [hunk 2]
```diff
@@ -1,2 +1,4 @@
 ^.gitignore$
 ^.travis.yml$
+^.*\.Rproj$
```
Comment (jameslamb) [inline .Rbuildignore:3]: thank you!

File: R/roxy_docsy.R [hunk 3]
```diff
@@ -1,3 +1,29 @@
...
+    # If class does not inherit anything, return NULL
+    if (is.null(aClass$get_inherit())) {
+        return(NULL)
+    }
+    
+    # Get parent class name and environment (package it's from)
+    parentClassName <- aClass$get_inherit()$classname
+    parentClassEnv <- environmentName(aClass$get_inherit()$parent_env)
+    
+    out <- sprintf("Inherits: \\code{\\link[%s]{%s}}\n"
```
Comment (jameslamb) [inline R/roxy_docsy.R:20]: niiiiiiice. Nothing like writing a little pseudo-LaTex with string interpolation!

File: R/roxy_docsy.R [hunk 4]
```diff
@@ -1,3 +1,29 @@
...
+# [description] Document inheritance and link to parent class
+.inherit_link <- function(aClass){
+    
+    # If class does not inherit anything, return NULL
+    if (is.null(aClass$get_inherit())) {
+        return(NULL)
+    }
+    
+    # Get parent class name and environment (package it's from)
+    parentClassName <- aClass$get_inherit()$classname
```
Comment (jameslamb) [inline R/roxy_docsy.R:17]: TIL `$get_inherit()`. Cool!

File: R/roxy_docsy.R [hunk 5]
```diff
@@ -38,7 +64,10 @@
     )
 
     # Add arguments?
-    los_argumentos <- names(formals(func))
+    
+    # Throws unavoidable warning if no arguments, but we don't really care
+    los_argumentos <- suppressWarnings(names(formals(func)))
```
Comment (jameslamb) [inline R/roxy_docsy.R]: this is a personal preference, but since it's a personal repo...can you change the formatting of this to

```
suppressWarnings({
    los_argumentos <- names(formals(func))
})
```

I like to be able to see the entire statement together instead of having to mentally skip over the `suppressWarnings()`

File: R/roxy_docsy.R [hunk 6]
```diff
@@ -76,6 +105,7 @@
     return(out)
 }
 
+# [description] Class constructor (initialize) block
```
Comment (jameslamb) [inline R/roxy_docsy.R:111]: thank you for adding these comments. Sorry they weren't there before

File: R/roxy_docsy.R [hunk 7]
```diff
@@ -86,64 +116,164 @@
...
+        out <- c(
+            out
+            , stats::setNames(
+                object = .describe_function(aClass$public_methods[[thisFuncName]], thisFuncName)
+                , nm = thisFuncName 
+            )
+        )
+    }
+    
+    # Identify duplicates by name. We want to keep the latest one because it will be a child class'
```
Comment (jameslamb) [inline R/roxy_docsy.R:157]: fantastic commenting!

File: R/roxy_docsy.R [hunk 8]
```diff
@@ -86,64 +116,164 @@
...
+# [description] Generate special methods block
+.special_methods_block <- function(aClass){
+    
+    descriptVec <- NULL
+    
+    # Iterate through this class' public methods
+    for (thisFuncName in SPECIAL_METHODS){
+        
+        # if not defined, skip it
+        if (!thisFuncName %in% names(aClass$public_methods)) {
```
Comment (jameslamb) [inline R/roxy_docsy.R:250]: when would this happen? Like if R6 changes the available set of baked-in things? (given that you hard-coded them into `SPECIAL_METHODS`)

If that's true, could you add a warning here saying "hey R6 has changed go change rsixygen"

File: R/roxy_docsy.R [hunk 9]
```diff
@@ -186,16 +316,18 @@ document_class <- function(aClass){
...
+    outSections <- c(
+        "\n "
+        , .inherit_link(aClass)
         , .constructor_block(aClass)
-        , "\n"
         , .public_methods_block(aClass)
-        , "\n"
-        , .public_member_block(aClass)
+        , .public_field_block(aClass)
+        , .special_methods_block(aClass)
```
Comment (jameslamb) [inline R/roxy_docsy.R:331]: I like the special methods block! great idea.

Although I do think it deserved to be it's own PR

File: R/roxy_docsy.R [hunk 10]
```diff
@@ -1,3 +1,29 @@
...
+.inherit_link <- function(aClass){
+    
+    # If class does not inherit anything, return NULL
+    if (is.null(aClass$get_inherit())) {
+        return(NULL)
+    }
+    
+    # Get parent class name and environment (package it's from)
+    parentClassName <- aClass$get_inherit()$classname
+    parentClassEnv <- environmentName(aClass$get_inherit()$parent_env)
```
Comment (jameslamb) [inline R/roxy_docsy.R:18]: wow cool!

**Review by jameslamb — CHANGES_REQUESTED** [review]
Well done, thanks Jay! I requested a few very minor things. Love this PR, this is really awesome.

File: R/roxy_docsy.R (hunk 8 above)
Comment (jayqi) [reply R/roxy_docsy.R:250]: This was in my PR notes but I can add a comment to the code too:

> Added a section "Special Methods" for special R6 methods like finalize, print, and clone. clone always exists, while the other two depend on people defining them.

**Comment by jayqi** [comment]
@jameslamb rebased and made the requested changes.

File: R/roxy_docsy.R (hunk 8 above)
Comment (jayqi) [reply R/roxy_docsy.R:250]: @jameslamb let me know if this is sufficient.

Comment (jameslamb) [reply R/roxy_docsy.R:250]: yeah this is perfect!

**Review by jameslamb — APPROVED** [review]
Looks awesome! Thanks for doing this Jay!!!
